File: e2e/go/scrape_test.go
// E2e tests for category: scrape
package e2e_test

import (
	"strings"
	"testing"

	pkg "github.com/kreuzberg-dev/kreuzcrawl"
)

func Test_ScrapeAssetDedup(t *testing.T) {
	// Same asset linked twice results in one download with one unique hash
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Assets) != 2 {
		t.Errorf("equals mismatch: got %v", len(result.Assets))
	}
	if len(result.Assets[0].ContentHash) == 0 {
		t.Errorf("expected non-empty value")
	}
}

func Test_ScrapeAssetMaxSize(t *testing.T) {
	// Skips assets exceeding max_asset_size limit
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Assets) != 2 {
		t.Errorf("equals mismatch: got %v", len(result.Assets))
	}
}

func Test_ScrapeAssetTypeFilter(t *testing.T) {
	// Only downloads image assets when asset_types filter is set
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Assets) != 1 {
		t.Errorf("equals mismatch: got %v", len(result.Assets))
	}
	if !strings.Contains(string(result.Assets[0].AssetCategory), `image`) {
		t.Errorf("expected to contain %s, got %v", `image`, result.Assets[0].AssetCategory)
	}
}

func Test_ScrapeBasicHtmlPage(t *testing.T) {
	// Scrapes a simple HTML page and extracts title, description, and links
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataTitle string
	if result.Metadata.Title != nil {
		metadataTitle = *result.Metadata.Title
	}
	var metadataDescription string
	if result.Metadata.Description != nil {
		metadataDescription = *result.Metadata.Description
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if strings.TrimSpace(result.ContentType) != `text/html` {
		t.Errorf("equals mismatch: got %v", result.ContentType)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(metadataTitle) != `Example Domain` {
		t.Errorf("equals mismatch: got %v", metadataTitle)
	}
	if !strings.Contains(string(metadataDescription), `illustrative examples`) {
		t.Errorf("expected to contain %s, got %v", `illustrative examples`, metadataDescription)
	}
	if result.Metadata.CanonicalUrl == nil || len(*result.Metadata.CanonicalUrl) == 0 {
		t.Errorf("expected non-empty value")
	}
	if len(result.Links) < 1 {
		t.Errorf("expected > 0, got %v", len(result.Links))
	}
	if !strings.Contains(string(result.Links[0].LinkType), `external`) {
		t.Errorf("expected to contain %s", `external`)
	}
	if len(result.Images) != 0 {
		t.Errorf("equals mismatch: got %v", len(result.Images))
	}
	if result.Metadata.OgTitle != nil && len(*result.Metadata.OgTitle) != 0 {
		t.Errorf("expected empty value, got %v", result.Metadata.OgTitle)
	}
}

func Test_ScrapeComplexLinks(t *testing.T) {
	// Classifies links by type: internal, external, anchor, document, image
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Links) < 10 {
		t.Errorf("expected > 9, got %v", len(result.Links))
	}
	if !strings.Contains(string(result.Links[0].LinkType), `internal`) {
		t.Errorf("expected to contain %s", `internal`)
	}
	if !strings.Contains(string(result.Links[0].LinkType), `external`) {
		t.Errorf("expected to contain %s", `external`)
	}
	if !strings.Contains(string(result.Links[0].LinkType), `anchor`) {
		t.Errorf("expected to contain %s", `anchor`)
	}
	if !strings.Contains(string(result.Links[0].LinkType), `document`) {
		t.Errorf("expected to contain %s", `document`)
	}
}

func Test_ScrapeDownloadAssets(t *testing.T) {
	// Downloads CSS, JS, and image assets from page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Assets) < 3 {
		t.Errorf("expected > 2, got %v", len(result.Assets))
	}
}

func Test_ScrapeDublinCore(t *testing.T) {
	// Extracts Dublin Core metadata from a page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataDcTitle string
	if result.Metadata.DcTitle != nil {
		metadataDcTitle = *result.Metadata.DcTitle
	}
	var metadataDcCreator string
	if result.Metadata.DcCreator != nil {
		metadataDcCreator = *result.Metadata.DcCreator
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(metadataDcTitle) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(metadataDcTitle) != `Effects of Climate Change on Marine Biodiversity` {
		t.Errorf("equals mismatch: got %v", metadataDcTitle)
	}
	if strings.TrimSpace(metadataDcCreator) != `Dr. Jane Smith` {
		t.Errorf("equals mismatch: got %v", metadataDcCreator)
	}
}

func Test_ScrapeEmptyPage(t *testing.T) {
	// Handles an empty HTML document without errors
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Links) <= -1 {
		t.Errorf("expected > -1, got %v", len(result.Links))
	}
	if len(result.Images) != 0 {
		t.Errorf("equals mismatch: got %v", len(result.Images))
	}
}

func Test_ScrapeFeedDiscovery(t *testing.T) {
	// Discovers RSS, Atom, and JSON feed links
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Feeds) < 3 {
		t.Errorf("expected >= 3, got %v", len(result.Feeds))
	}
}

func Test_ScrapeImageSources(t *testing.T) {
	// Extracts images from img, picture, og:image, twitter:image
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataOgImage string
	if result.Metadata.OgImage != nil {
		metadataOgImage = *result.Metadata.OgImage
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Images) < 5 {
		t.Errorf("expected > 4, got %v", len(result.Images))
	}
	if strings.TrimSpace(metadataOgImage) != `https://example.com/images/og-hero.jpg` {
		t.Errorf("equals mismatch: got %v", metadataOgImage)
	}
}

func Test_ScrapeJsHeavySpa(t *testing.T) {
	// Handles SPA page with JavaScript-only content (no server-rendered HTML)
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
}

func Test_ScrapeJsonLd(t *testing.T) {
	// Extracts JSON-LD structured data from a page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var jsonLdName string
	if result.JsonLd[0].Name != nil {
		jsonLdName = *result.JsonLd[0].Name
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.JsonLd) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(result.JsonLd[0].SchemaType) != `Recipe` {
		t.Errorf("equals mismatch: got %v", result.JsonLd[0].SchemaType)
	}
	if strings.TrimSpace(jsonLdName) != `Best Chocolate Cake` {
		t.Errorf("equals mismatch: got %v", jsonLdName)
	}
}

func Test_ScrapeMalformedHtml(t *testing.T) {
	// Gracefully handles broken HTML without crashing
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataDescription string
	if result.Metadata.Description != nil {
		metadataDescription = *result.Metadata.Description
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(result.Html) == 0 {
		t.Errorf("expected non-empty value")
	}
	if !strings.Contains(string(metadataDescription), `broken HTML`) {
		t.Errorf("expected to contain %s, got %v", `broken HTML`, metadataDescription)
	}
}

func Test_ScrapeOgMetadata(t *testing.T) {
	// Extracts full Open Graph metadata from a page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataOgTitle string
	if result.Metadata.OgTitle != nil {
		metadataOgTitle = *result.Metadata.OgTitle
	}
	var metadataOgType string
	if result.Metadata.OgType != nil {
		metadataOgType = *result.Metadata.OgType
	}
	var metadataOgImage string
	if result.Metadata.OgImage != nil {
		metadataOgImage = *result.Metadata.OgImage
	}
	var metadataTitle string
	if result.Metadata.Title != nil {
		metadataTitle = *result.Metadata.Title
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if len(metadataOgTitle) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(metadataOgTitle) != `Article Title` {
		t.Errorf("equals mismatch: got %v", metadataOgTitle)
	}
	if strings.TrimSpace(metadataOgType) != `article` {
		t.Errorf("equals mismatch: got %v", metadataOgType)
	}
	if strings.TrimSpace(metadataOgImage) != `https://example.com/images/article-hero.jpg` {
		t.Errorf("equals mismatch: got %v", metadataOgImage)
	}
	if result.Metadata.OgDescription == nil || len(*result.Metadata.OgDescription) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(metadataTitle) != `Article Title - Example Blog` {
		t.Errorf("equals mismatch: got %v", metadataTitle)
	}
}

func Test_ScrapeTwitterCard(t *testing.T) {
	// Extracts Twitter Card metadata from a page
	engine, createErr := pkg.CreateEngine()
	if createErr != nil {
		t.Fatalf("create handle failed: %v", createErr)
	}
	result, err := pkg.Scrape(engine, "")
	if err != nil {
		t.Fatalf("call failed: %v", err)
	}
	var metadataTwitterCard string
	if result.Metadata.TwitterCard != nil {
		metadataTwitterCard = *result.Metadata.TwitterCard
	}
	var metadataTwitterTitle string
	if result.Metadata.TwitterTitle != nil {
		metadataTwitterTitle = *result.Metadata.TwitterTitle
	}
	if result.StatusCode != 200 {
		t.Errorf("equals mismatch: got %v", result.StatusCode)
	}
	if result.Metadata.TwitterCard == nil || len(*result.Metadata.TwitterCard) == 0 {
		t.Errorf("expected non-empty value")
	}
	if strings.TrimSpace(metadataTwitterCard) != `summary_large_image` {
		t.Errorf("equals mismatch: got %v", metadataTwitterCard)
	}
	if strings.TrimSpace(metadataTwitterTitle) != `New Product Launch` {
		t.Errorf("equals mismatch: got %v", metadataTwitterTitle)
	}
}

File: e2e/elixir/test/scrape_test.exs
# E2e tests for category: scrape
defmodule E2e.ScrapeTest do
  use ExUnit.Case, async: true

  describe "scrape_asset_dedup" do
    test "Same asset linked twice results in one download with one unique hash" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.assets)) == 2
      assert result.assets[0].content_hash != ""
    end
  end

  describe "scrape_asset_max_size" do
    test "Skips assets exceeding max_asset_size limit" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.assets)) == 2
    end
  end

  describe "scrape_asset_type_filter" do
    test "Only downloads image assets when asset_types filter is set" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert String.trim(length(result.assets)) == 1
      assert String.contains?(result.assets[0].asset_category, "image")
    end
  end

  describe "scrape_basic_html_page" do
    test "Scrapes a simple HTML page and extracts title, description, and links" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert String.trim(result.content_type) == "text/html"
      assert result.html != ""
      assert String.trim(result.metadata.title) == "Example Domain"
      assert String.contains?(result.metadata.description, "illustrative examples")
      assert result.metadata.canonical_url != ""
      assert length(result.links) > 0
      assert String.contains?(result.links[0].link_type, "external")
      assert String.trim(length(result.images)) == 0
      assert String.trim(result.metadata.og_title) == ""
    end
  end

  describe "scrape_complex_links" do
    test "Classifies links by type: internal, external, anchor, document, image" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert length(result.links) > 9
      assert String.contains?(result.links[0].link_type, "internal")
      assert String.contains?(result.links[0].link_type, "external")
      assert String.contains?(result.links[0].link_type, "anchor")
      assert String.contains?(result.links[0].link_type, "document")
    end
  end

  describe "scrape_download_assets" do
    test "Downloads CSS, JS, and image assets from page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert length(result.assets) > 2
    end
  end

  describe "scrape_dublin_core" do
    test "Extracts Dublin Core metadata from a page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert result.metadata.dc_title != ""
      assert String.trim(result.metadata.dc_title) == "Effects of Climate Change on Marine Biodiversity"
      assert String.trim(result.metadata.dc_creator) == "Dr. Jane Smith"
    end
  end

  describe "scrape_empty_page" do
    test "Handles an empty HTML document without errors" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert length(result.links) > -1
      assert String.trim(length(result.images)) == 0
    end
  end

  describe "scrape_feed_discovery" do
    test "Discovers RSS, Atom, and JSON feed links" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert length(result.feeds) >= 3
    end
  end

  describe "scrape_image_sources" do
    test "Extracts images from img, picture, og:image, twitter:image" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert length(result.images) > 4
      assert String.trim(result.metadata.og_image) == "https://example.com/images/og-hero.jpg"
    end
  end

  describe "scrape_js_heavy_spa" do
    test "Handles SPA page with JavaScript-only content (no server-rendered HTML)" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert result.html != ""
    end
  end

  describe "scrape_json_ld" do
    test "Extracts JSON-LD structured data from a page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert result.json_ld != ""
      assert String.trim(result.json_ld[0].schema_type) == "Recipe"
      assert String.trim(result.json_ld[0].name) == "Best Chocolate Cake"
    end
  end

  describe "scrape_malformed_html" do
    test "Gracefully handles broken HTML without crashing" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert result.html != ""
      assert String.contains?(result.metadata.description, "broken HTML")
    end
  end

  describe "scrape_og_metadata" do
    test "Extracts full Open Graph metadata from a page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert result.metadata.og_title != ""
      assert String.trim(result.metadata.og_title) == "Article Title"
      assert String.trim(result.metadata.og_type) == "article"
      assert String.trim(result.metadata.og_image) == "https://example.com/images/article-hero.jpg"
      assert result.metadata.og_description != ""
      assert String.trim(result.metadata.title) == "Article Title - Example Blog"
    end
  end

  describe "scrape_twitter_card" do
    test "Extracts Twitter Card metadata from a page" do
      engine = Kreuzcrawl.create_engine!(nil)
      result = Kreuzcrawl.scrape!(engine, "")
      assert String.trim(result.status_code) == 200
      assert result.metadata.twitter_card != ""
      assert String.trim(result.metadata.twitter_card) == "summary_large_image"
      assert String.trim(result.metadata.twitter_title) == "New Product Launch"
    end
  end
end
